trieste/models/gpflux/models.py

```diff
@@ -0,0 +1,158 @@
+                "`LatentVariableLayer`"
+            )
+
+        self._model_gpflux = model
+
+        self._model_keras = model.as_training_model()
+        self._model_keras.compile(self._optimizer)
+
+    def __repr__(self) -> str:
+        """"""
+        return f"DeepGaussianProcess({self._model_gpflux!r}, {self.optimizer!r})"
+
+    @property
+    def model_gpflux(self) -> DeepGP:
+        return self._model_gpflux
+
+    @property
+    def model_keras(self) -> tf.keras.Model:
+        return self._model_keras
+
+    @property
+    def optimizer(self) -> tf.keras.optimizers.Optimizer:
+        return self._optimizer
+
+    def sample(self, query_points: TensorType, num_samples: int) -> TensorType:
+        def get_samples(query_points: TensorType, num_samples: int) -> TensorType:
+            samples = []
+            for _ in range(num_samples):
+                samples.append(sample_dgp(self.model_gpflux)(query_points))
+            return tf.stack(samples)
+
+        return get_samples(query_points, num_samples)
+
+    def update(self, dataset: Dataset) -> None:
+        inputs = dataset.query_points
+        new_num_data = inputs.shape[0]
+        self.model_gpflux.num_data = new_num_data
+
+        # Update num_data for each layer, as well as make sure dataset shapes are ok
+        for i, layer in enumerate(self.model_gpflux.f_layers):
+            layer.num_data = new_num_data
+
+            if isinstance(layer, LatentVariableLayer):
+                inputs = layer(inputs)
+                continue
+
+            if isinstance(layer.inducing_variable, InducingPoints):
+                inducing_variable = layer.inducing_variable
+            else:
+                inducing_variable = layer.inducing_variable.inducing_variable
+
+            if inputs.shape[-1] != inducing_variable.Z.shape[-1]:
+                raise ValueError(
+                    f"Shape {inputs.shape} of input to layer {layer} is incompatible with shape"
+                    f" {inducing_variable.Z.shape} of that layer. Trailing dimensions must match."
+                )
+
+            if i == len(self.model_gpflux.f_layers) - 1:
+                if dataset.observations.shape[-1] != layer.q_mu.shape[-1]:
+                    raise ValueError(
+                        f"Shape {dataset.observations.shape} of new observations is incompatible"
+                        f" with shape {layer.q_mu.shape} of existing observations. Trailing"
+                        f" dimensions must match."
+                    )
+
+            inputs = layer(inputs)
+
+    def optimize(self, dataset: Dataset) -> None:
+        """
+        Optimize the model with the specified `dataset`.
+        :param dataset: The data with which to optimize the `model`.
+        """
+        self.model_keras.fit(
+            {"inputs": dataset.query_points, "targets": dataset.observations}, **self.fit_args
+        )
+
+        # Reset lr in case there was an lr schedule
```

Comment:
Can you elaborate a bit more in the comment? This seems to be talking about some obscure tf/keras mechanics